zilliqa/src/consensus.rs

```diff
@@ -472,87 +548,111 @@ impl Consensus {
 
-            // If it is higher, mark it for a retry
-            if tx.transaction.nonce > nonce {
-                let retries = self.new_transactions_waiting.entry(tx.hash()).or_insert(0);
-                let _ = retries.add(1);
-                warn!(
-                        "Transaction nonce for tx {} is too high: {} when acct nonce is {}, tries: {}/{}",
-                        tx.hash(),
-                        tx.transaction.nonce,
-                        self.state.must_get_account(tx.from_addr).nonce,
-                        retries,
+                // if it is less than the iter nonce, we can skip it
+                // this just means there are 'priority txs' with the same nonce but different fees
+                if txn.nonce < iter_nonce {
+                    trace!("lower priority tx ignored {} vs {}", txn.nonce, iter_nonce);
+                    continue;
+                }
+
+                // If the nonce is too high, we will take another pass
+                if txn.nonce > iter_nonce {
+                    trace!(
+                        "TX nonce {} too high, iter nonce {}, retry {} of {}",
+                        txn.nonce,
+                        iter_nonce,
+                        txn.retries,
                         self.config.tx_retries
                     );
 
-                if *retries >= self.config.tx_retries {
-                    warn!(?tx, "Tranaction has exceeded retries and will been removed");
-                    tx_hashes_invalid.push(tx.hash());
-                    continue;
+                    if txn.retries >= self.config.tx_retries {
+                        warn!(?txn, "Tranaction has exceeded retries and all pending from this account will been removed");
+                        remove_all_txs = true;
+                    }
+
+                    // all other TXs will have a higher nonce, so we can break
+                    break;
                 }
-            }
 
-            // Tx nonce is good to add, we need to check that there isn't already a tx with the same
-            // from address, though (no need to drop though)
-            if tx_from_addrs.contains(&tx.from_addr) {
-                warn!(
-                    "Transaction from address {} already exists in block and will not be added",
-                    tx.from_addr
-                );
-                continue;
+                // tx nonce == iter nonce, so we can add it
+                match self.new_transactions.get(&txn.hash) {
+                    Some(tx) => {
+                        trace!("Adding tx to execute: {:?}", tx);
+                        ret.push(tx.clone());
+                        iter_nonce += 1;
+                    }
+                    None => {
+                        warn!(
+                            "Transaction {} not found in mempool, but was in priority txs",
+                            txn.hash
+                        );
+                        remove_all_txs = true;
```

Comment:
What's the reason for clearing out the entire priority map if one inconsistent header is found? Having the data be inconsistent sounds like it would be symptom of a bug - I'd probably flag it as an error condition, then either ignore it hoping it's a one off or rebuild the entire heap by scanning `new_transactions`. I'm a bit confused at the choice here

Reply:
Yes, that is a good point - ok, I will just drop that TX from temp_vec and have the warning